src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersCodeFixProvider.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.RemoveUnusedMembers
+{
+    internal abstract class AbstractRemoveUnusedMembersCodeFixProvider<TFieldDeclarationSyntax> : SyntaxEditorBasedCodeFixProvider
+        where TFieldDeclarationSyntax : SyntaxNode
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.RemoveUnusedMembersDiagnosticId);
+
+        /// <summary>
+        /// This method adjusts the <paramref name="declarators"/> to remove based on whether or not all variable declarators
+        /// within a field declaration should be removed,
+        /// i.e. if all the fields declared within a field declaration are unused,
+        /// we can remove the entire field declaration instead of individual variable declarators.
+        /// </summary>
+        protected abstract void AdjustAndAddAppropriateDeclaratorsToRemove(HashSet<TFieldDeclarationSyntax> fieldDeclarators, HashSet<SyntaxNode> declarators);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(new MyCodeAction(
+                c => FixAsync(context.Document, context.Diagnostics[0], c)),
+                context.Diagnostics);
+            return Task.CompletedTask;
+        }
+
+        protected override async Task FixAllAsync(
+            Document document,
+            ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor,
+            CancellationToken cancellationToken)
+        {
+            var declarators = new HashSet<SyntaxNode>();
+            var fieldDeclarators = new HashSet<TFieldDeclarationSyntax>();
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declarationService = document.GetLanguageService<ISymbolDeclarationService>();
+
+            // Compute declarators to remove, and also track common field declarators.
+            foreach (var diagnostic in diagnostics)
+            {   
+                // Get symbol to be removed.
+                var diagnosticNode = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
+                var symbol = semanticModel.GetDeclaredSymbol(diagnosticNode, cancellationToken);
```

Comment:
note: that this works is *super* fragile.  You are using the Location of the diagnostic (which is often a token of the member).  You're then calling FindNode, which *fortunately* is always finding the node that works for GetDeclaredSymbol.

What would be *safer* is to use .AdditionalLocations to store the actual Symbol *Location[0]*.

Reply:
Diagnostic is reported with the primary location as `Symbol.Locations[0]`.

---
In reply to: [213882742](https://github.com/dotnet/roslyn/pull/29511#discussion_r213882742) [](ancestors = 213882742)